### PR TITLE
Extend the `#[must_use]` lint to arrays

### PR DESCRIPTION
**Pull request by varkor**
Based on top of https://github.com/rust-lang/rust/pull/62228.

r? @Centril

### CODE REVIEW

File: src/librustc_lint/unused.rs [hunk 1]
```diff
@@ -179,14 +198,27 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnusedResults {
...
                     }
                     has_emitted
                 }
+                ty::Array(ty, len) => match len.assert_usize(cx.tcx) {
+                    // If the array is definitely non-empty, we can do `#[must_use]` checking.
+                    Some(n) if n != 0 => {
+                        let descr_pre = &format!(
+                            "{}array{} of ",
+                            descr_pre,
+                            if plural { "s" } else { "" },
```
Comment (Centril) [inline src/librustc_lint/unused.rs]: Third time this is used -- deduplicate as a closure or something.

File: src/librustc_lint/unused.rs [hunk 2]
```diff
@@ -179,14 +198,27 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnusedResults {
...
+                        let descr_post = &format!(" in tuple element {}", i);
                         let span = *spans.get(i).unwrap_or(&span);
-                        if check_must_use_ty(cx, ty, expr, span, descr_post_path) {
+                        if check_must_use_ty(cx, ty, expr, span, descr_pre, descr_post, plural) {
                             has_emitted = true;
                         }
                     }
                     has_emitted
                 }
+                ty::Array(ty, len) => match len.assert_usize(cx.tcx) {
```
Comment (Centril) [inline src/librustc_lint/unused.rs]: Could possibly be nicer as `ty::Array(ty, len) if len.assert_usize(cx.tcx).filter(|&n| n > 0).is_some() => { ... }`.
Up to you :)

**Comment by Centril** [comment]
r=me with CI passing, the first nit fixed, and optionally the second.

@rust-lang/lang N.B. This PR extends `#[must_use]` to arrays. I believe this should be uncontroversial given https://github.com/rust-lang/rust/pull/61100 and our approval of that given that an array can be viewed as a sort of homogeneous tuple type. Arrays are also structural-ish.

**Comment by varkor** [comment]
This will be fixed once https://github.com/rust-lang/miri/pull/806 is merged.

**Comment by Centril** [comment]
r=me then :)

**Pull request merged**